iroha/src/kura.rs

```diff
@@ -180,126 +196,248 @@ impl Default for Mode {
     #[iroha_futures::telemetry_future]
     pub async fn write(&self, block: &VersionedCommittedBlock) -> Result<Hash> {
-        //filename is its height
-        let path = self.get_block_path(block.header().height);
-        let mut file = File::create(path)
+        let path = self
+            .get_block_path(
+                NonZeroU64::new(block.header().height)
+                    .wrap_err("An attempt to write a zero-height block.")?,
+            )
+            .await?;
+        let mut file = fs::OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(path)
             .await
             .wrap_err("Failed to open storage file.")?;
         let hash = block.hash();
         let serialized_block: Vec<u8> = block.encode_versioned()?;
+        #[allow(clippy::cast_possible_truncation)]
+        let block_size = serialized_block.len() as u32;
+        file.write_u32_le(block_size).await.wrap_err("Failed to write to storage file.")?;
         file.write_all(&serialized_block)
             .await
             .wrap_err("Failed to write to storage file.")?;
         Ok(hash)
     }
 
-    /// Returns a block at height, read from disk.
-    ///
-    /// # Errors
-    /// - There is no file corresponding to this block.
+    /// Returns a sorted vector of blocks starting from 1 height to the top block.
     #[iroha_futures::telemetry_future]
-    pub async fn read(&self, height: u64) -> Result<VersionedCommittedBlock> {
-        let path = self.get_block_path(height);
-        let mut file = File::open(&path).await.wrap_err("No file found.")?;
-        let metadata = metadata(&path).await.wrap_err("Unable to read metadata.")?;
-        #[allow(clippy::cast_possible_truncation)]
-        let mut buffer = vec![0; metadata.len() as usize];
-        let _ = file.read(&mut buffer).await.wrap_err("Buffer overflow.")?;
-        VersionedCommittedBlock::decode_versioned(&buffer)
-            .wrap_err("Failed to read block from store.")
+    pub async fn read_all(&self) -> Result<Vec<VersionedCommittedBlock>> {
+        let mut blocks: Vec<VersionedCommittedBlock> = Vec::new();
+        let base_indices = storage_files_base_indices(&self.path).await?;
+        for base_index in base_indices {
+            let path = self.path.join(base_index.to_string());
+            let mut read_blocks = read_all_from_storage_file(&path)
+                .await
+                .wrap_err("Failed to deserialize blocks.")?;
+            let prev_block = read_blocks
+                .first()
+                .wrap_err(format!(
+                    "No blocks were found in file {}. Data storage might be corrupted.",
+                    path.to_string_lossy()
+                ))?
+                .header()
+                .height;
+            if let Some(last_block) = blocks.last() {
+                let new_block = last_block.header().height;
+                if prev_block != new_block + 1 {
```

Comment:
This whole block is hairy at the moment.
I tried my best to explode everything I do type-by-type, and even went from map().collect() to for_each(forward()), but still cannot untangle how it should be done properly.

Reply:
Also, for the time being I ignore error handling - it's a thing to account for as well.